xapian-core/include/xapian/cluster.h

```diff
@@ -0,0 +1,412 @@
+    virtual ~Similarity();
+
+    /// Calculates the similarity between the two documents
+    virtual double similarity(PointType &a, PointType &b) const = 0;
+
+    /// Returns description of the similarity metric being used
+    virtual std::string get_description() const = 0;
+};
+
+/** Class for calculating the cosine distance between two documents
+ */
+class XAPIAN_VISIBILITY_DEFAULT CosineDistance : public Similarity {
+
+  public:
+
+    /** Calculates and returns the cosine similarity using the
+     *  formula  cos(theta) = a.b/(|a|*|b|)
+     */
+    double similarity(PointType &a, PointType &b) const;
+
+    /// Returns the description of Cosine Similarity
+    std::string get_description() const;
+};
+
+/** Class representing an abstract class for a clusterer to be implemented
+ */
+class XAPIAN_VISIBILITY_DEFAULT Clusterer {
+
+  public:
+
+    /// Destructor
+    virtual ~Clusterer();
+
+    /* Implement the required clustering algorithm in the subclass and
+     *  and return clustered output as ClusterSet
+     */
+    virtual ClusterSet cluster(MSet &mset) = 0;
+
+    /// Returns a description of the clusterer being used
+    virtual std::string get_description() const = 0;
+};
+
+/** Round Robin clusterer:
+ *  This clusterer is a minimal clusterer which will cluster documents as -
+ *  ith document goes to the (i % k)th cluster where k is the number of clusters and
+ *  0 <= i < N; where N is the number of documents
+ */
+class XAPIAN_VISIBILITY_DEFAULT RoundRobin : public Clusterer {
+
+    /// Number of clusters to be formed by the clusterer
+    unsigned int num_of_clusters;
+
+  public:
+
+    /// Constructor
+    RoundRobin(unsigned int num_of_clusters_) : num_of_clusters(num_of_clusters_) {}
+
+    /// Implements the RoundRobin clustering
+    ClusterSet cluster(MSet &mset);
+
+    /// Returns the description of the clusterer
+    std::string get_description() const;
+};
+
+/** Kmeans clusterer:
+ *  This clusterer implements the K-Means clustering algorithm
+ */
+class XAPIAN_VISIBILITY_DEFAULT KMeans : public Clusterer {
+
+    /// Contains the initialized points that are to be clustered
+    std::vector<Point> docs;
+
+    /// Specifies that the clusterer needs to form 'k' clusters
+    unsigned int k;
+
+    /// Specifies the maximum number of iterations that KMeans will have
+    unsigned int max_iters;
+
+    /** Initialize 'k' clusters by randomly selecting 'k' centroids
+     *  and assigning them to different clusters
+     */
+    void initialize_clusters(ClusterSet &cset);
+
+    /** Initialize the 'Points' to be fed into the Clusterer with the DocumentSource.
+     *  The TF-IDF weights for the points are calculated and stored within the
+     *  Points to be used later during distance calculations
+     */
+    void initialize_points(const MSet &docs);
+
+  public:
+
+    /// Constructor specifying number of clusters and maximum iterations
+    KMeans(unsigned int k_, unsigned int max_iters_ = MAX_ITERS);
```

Comment:
This constructor has a single argument form, so should be marked as `explicit` (we don't want the compiler to implicitly convert `unsigned int` to `KMeans`).